### PR TITLE
DRAFT: Feat/contract test pipeline

### PR DESCRIPTION
**Pull request by JackMidd-Amido**
https://amido-dev.visualstudio.com/Amido-Stacks/_workitems/edit/8412

#### 📲 What

Added Contract Testing to the simple-api/webapi template.
These contract tests do the following:
- publish the OpenAPI specification to PactFlow
- run `can-i-deploy` before running any deployment steps
- Record deployment of the application to each environment in PactFlow

#### 🤔 Why

This allows us to support Bi-Directional contract testing out of the box.
This will significantly decrease the time to set up contract testing on projects

#### 🛠 How

Utilising the PactFlow docker container and existing pipelines.

#### 👀 Evidence

https://dev.azure.com/ensonodigitaluk/Stacks/_build/results?buildId=6868&view=results

#### 🕵️ How to test

Run the pipeline above and check that PactFlow has been updated accordingly:
https://ensono-stacks.pactflow.io/pacticipants/stacks-provider/versions/7.0.709

#### ✅ Acceptance criteria Checklist

- [ ] Code peer reviewed?
- [ ] Documentation has been updated to reflect the changes?
- [ ] Passing all automated tests, including a successful deployment?
- [ ] Passing any exploratory testing?
- [ ] Rebased/merged with latest changes from development and re-tested?
- [ ] Meeting the Coding Standards?

### CODE REVIEW

**Review by saulfrance — CHANGES_REQUESTED** [review]
I believe you will also need to make similar changes for cqrs and the background worker. When generated using the dotnet template they all use the same `build/azDevOps/azure/ci.yml` file.

Could you make a flag like `run_pact_tests` so that it is optional?

Do we also need to apply this to the GitHub Actions pipelines?

**Comment by JackMidd-Amido** [comment]
@saulfrance - I've updated cqrs and background-worker pipelines now too.

I don't think these need to be added to the github actions pipelines as these tests aren't actually testing anything on their own, out of the box. 
We are adding them to the templates so that users have the option of implementing contract tests quickly when they create a project using the templates.

Does that make sense? Happy to discuss further, I may have misunderstood the purpose of the actions pipelines

File: build/azDevOps/azure/ci.yml [hunk 1]
```diff
@@ -131,7 +166,32 @@ stages:
...
+                  -e PACT_BROKER_BASE_URL=$(PACT_BROKER_BASE_URL) \
+                  -e PACT_BROKER_TOKEN=$(PACT_BROKER_TOKEN) \
+                  pactfoundation/pact-cli:latest \
+                  broker can-i-deploy \
+                  --pacticipant $(PACTICIPANT_NAME) \
+                  --version $(version_number) \
+                  --to-environment $(Environment.ShortName)
+
       - deployment: AppInfraDev
+        dependsOn: canideploy_dev
```
Comment (saulfrance) [inline build/azDevOps/azure/ci.yml:194]: Is this needed? It's not in the the other pipeline files?

Comment (saulfrance) [reply build/azDevOps/azure/ci.yml:194]: If it is can you test that it still runs when the `run_contract_tests` flag is false?

Comment (JackMidd-Amido) [reply build/azDevOps/azure/ci.yml:194]: Ah, forgot about that, thanks. 

Will test it out, might need some custom condition here...

Comment (JackMidd-Amido) [reply build/azDevOps/azure/ci.yml]: I've changed this to a condition and added it on other deploy jobs.

**Pull request merged**